plugin/pkg/scheduler/schedulercache/cache.go

```diff
@@ -0,0 +1,268 @@
+}
+
+func (cache *schedulerCache) List(selector labels.Selector) ([]*api.Pod, error) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	var pods []*api.Pod
+	for _, info := range cache.nodes {
+		for _, pod := range info.pods {
+			if selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (cache *schedulerCache) AssumePodIfBindSucceed(pod *api.Pod, bind func() bool) error {
+	return cache.assumePodIfBindSucceed(pod, bind, time.Now())
+}
+
+// assumePodScheduled exists for making test deterministic by taking time as input argument.
+func (cache *schedulerCache) assumePodIfBindSucceed(pod *api.Pod, bind func() bool, now time.Time) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	if !bind() {
+		return nil
+	}
+
+	key, err := getPodKey(pod)
+	if err != nil {
+		return err
+	}
+	if _, ok := cache.podStates[key]; ok {
+		return fmt.Errorf("pod state wasn't initial but get assumed. Pod key: %v", key)
+	}
+
+	cache.addPod(pod)
+	ps := &podState{
+		pod:      pod,
+		deadline: now.Add(cache.ttl),
+	}
+	cache.podStates[key] = ps
+	cache.assumedPods[key] = true
+	return nil
+}
+
+func (cache *schedulerCache) AddPod(pod *api.Pod) error {
+	key, err := getPodKey(pod)
+	if err != nil {
+		return err
+	}
+
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	_, ok := cache.podStates[key]
+	switch {
+	case ok && cache.assumedPods[key]:
+		delete(cache.assumedPods, key)
+	case !ok:
+		// Pod was expired. We should add it back.
+		cache.addPod(pod)
+	default:
+		return fmt.Errorf("pod state wasn't assumed or expired but get added. Pod key: %v", key)
+	}
+	return nil
+}
+
+func (cache *schedulerCache) UpdatePod(oldPod, newPod *api.Pod) error {
+	key, err := getPodKey(oldPod)
+	if err != nil {
+		return err
+	}
+
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	_, ok := cache.podStates[key]
+	switch {
+	case ok && !cache.assumedPods[key]:
+		err := cache.updatePod(oldPod, newPod)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("pod state wasn't added but get updated. Pod key: %v", key)
+	}
+	return nil
+}
+
+func (cache *schedulerCache) updatePod(oldPod, newPod *api.Pod) error {
+	err := cache.deletePod(oldPod)
+	if err != nil {
+		return err
+	}
+	cache.addPod(newPod)
+	return nil
+}
+
+func (cache *schedulerCache) addPod(pod *api.Pod) {
+	n, ok := cache.nodes[pod.Spec.NodeName]
+	if !ok {
+		n = NewNodeInfo()
+		cache.nodes[pod.Spec.NodeName] = n
+	}
+	n.addPod(pod)
+}
+
+func (cache *schedulerCache) deletePod(pod *api.Pod) error {
+	n := cache.nodes[pod.Spec.NodeName]
+	err := n.removePod(pod)
+	if err != nil {
+		return err
+	}
+	if len(n.pods) == 0 {
+		delete(cache.nodes, pod.Spec.NodeName)
+	}
+	return nil
+}
+
+func (cache *schedulerCache) RemovePod(pod *api.Pod) error {
+	key, err := getPodKey(pod)
+	if err != nil {
+		return err
+	}
+
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	_, ok := cache.podStates[key]
+	switch {
+	case ok && !cache.assumedPods[key]:
```

Comment:
ahh - looking into your state machine it seems that I'm wrong.
So the pattern you would like to have is that in case of assumed pod, we simply ignore the Remove event, and wait for its expiration?
Can you please add a comment about it (so that noone will think that it's a bug in the future)?